File: c++/0876.cpp
// Link: https://leetcode.com/problems/middle-of-the-linked-list/description/
// Runtime: 0 ms
// Memory: 8.66 MB

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution 
{
public:
    ListNode* middleNode(ListNode* head) 
    {
        ListNode* root = head;
        std::vector<ListNode*> nodes;
        while (root != nullptr)
        {
            nodes.push_back(root);
            root = root->next;
        }

        return nodes[nodes.size() / 2];
    }
};

File: python/0876.py
# Link: https://leetcode.com/problems/middle-of-the-linked-list/description/
# Runtime: 32 ms
# Memory: 16.53 MB

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None or head.next == None:
            return head
        
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        return slow
